### PR TITLE
feat: multiple sharding keys implementation

### LINKED ISSUE #425: [Sharding] Support multiple columns complex sharding key.
**What would you like to be added**:

**Why is this needed**:

### PR DESCRIPTION
**Pull request by jjeffcaii**
**What this PR does**:
1. tow low-level modules includes `logic` and `calculus`.
2. sharding columns configuration refinement.

**Which issue(s) this PR fixes**:

Fixes #425 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by maronghe — APPROVED** [review]

**Review by maronghe — APPROVED** [review]

File: conf/config.yaml [hunk 1]
```diff
@@ -63,20 +63,39 @@ data:
...
               sqlMaxLimit: -1
+          - name: employees.friendship
+            sequence:
+              type: snowflake
+              option:
+            db_rules:
+              - columns:
+                  - name: uid
+                  - name: friend_id
+                expr: parseInt(($0*31+$1) % 32 / 8)
```
[the remark this answers is not in the document]
Comment (jjeffcaii) [reply conf/config.yaml:87]: The `$index` means the column of target index, and `$0*31+$1` is equals with `valueOf(uid) * 31 + valueOf(friend_id)`. BTW, it's just an example which show how the new expression works.

File: integration_test/config/db_tbl/config.yaml [hunk 2]
```diff
@@ -60,12 +60,12 @@ data:
               type: snowflake
               option:
             db_rules:
-              - column: uid
-                type: scriptExpr
+              - columns:
+                  - name: uid
                 expr: parseInt($value % 32 / 8)
```
[the remark this answers is not in the document]
Comment (jjeffcaii) [reply integration_test/config/db_tbl/config.yaml:65]: To be compatible with the old expression design, `$value` is the alias of `$0`.

File: pkg/proto/rule/rule.go [hunk 3]
```diff
@@ -115,44 +159,44 @@ func (vt *VTable) AllowFullScan() bool {
...
-	res["db_rules"] = string(val)
+	res["name"] = vt.Name()
+	res["sequence_type"] = vt.GetAutoIncrement().Type
 
-	val, err = json.Marshal(vt.rawShards.TblRules)
-	if err != nil {
-		return res, err
-	}
-	res["tbl_rules"] = string(val)
+	res["db_rules"] = "todo"
```
[the remark this answers is not in the document]
Comment (jjeffcaii) [reply pkg/proto/rule/rule.go:195]: Got it, I'll check it.

File: pkg/runtime/ast/create_table.go [hunk 4]
```diff
@@ -21,13 +21,11 @@ import (
...
 
-import (
-	"github.com/pkg/errors"
-)
-
 import (
 	"github.com/arana-db/parser/ast"
 	"github.com/arana-db/parser/format"
+
+	"github.com/pkg/errors"
```
[the remark this answers is not in the document]
Comment (jjeffcaii) [reply pkg/runtime/ast/create_table.go:28]: 👌🏽

File: pkg/proto/rule/topology.go [hunk 5]
```diff
@@ -104,6 +108,16 @@ func (to *Topology) Enumerate() DatabaseTables {
 	return dt
 }
 
+func (to *Topology) Exists(dbIdx, tbIdx int) bool {
```
Comment (dongzl) [inline pkg/proto/rule/topology.go:111]: why not `Exist`?

**Review by dongzl — APPROVED** [review]

**Comment by maronghe** [comment]
done

**Pull request merged**